lib/srv/db/server.go

```diff
@@ -1021,7 +1021,7 @@ func (s *Server) handleConnection(ctx context.Context, clientConn net.Conn) erro
 	if err != nil {
 		connectionDiagnosticID := sessionCtx.Identity.ConnectionDiagnosticID
 		if connectionDiagnosticID != "" && trace.IsAccessDenied(err) {
-			_, diagErr := s.cfg.AuthClient.AppendDiagnosticTrace(ctx,
+			_, diagErr := s.cfg.AuthClient.AppendDiagnosticTrace(cancelCtx,
```

Comment:
This change feels weird.
Diagnostics are used by the UI when adding/testing new nodes and testing the connection to them so we can loose them in any case as they are not an important part of teleport.

Second, they are blocking and will block until the append finishes. Since the cancelCtx is derived from ctx with a simple cancel function that executes on exit (defer cancel() on line 960), ctx and cancelCtx behave exactly the same here.

Reply:
`ctx` is later overwritten by `MonitorConn`.

Reply:
> `ctx` is later overwritten by `MonitorConn`.
Yes, exactly this.

> Diagnostics are used by the UI when adding/testing new nodes and testing the connection to them so we can loose them in any case as they are not an important part of teleport.

This is indeed optional and not a huge deal if this happens. But that would mean the check fails in the UI, and the flow is broken. I believe this is a reason enough to fix it.

Reply:
> do we care that the diagnostic trace doesn't get uploaded?

Yes, we do:

> This is indeed optional and not a huge deal if this happens. But that would mean the check fails in the UI, and the flow is broken. I believe this is a reason enough to fix it.

> Is `ctx` a request scoped context? Is it only cancelled by a user? 

The `ctx` which is currently passed to `AppendDiagnosticTrace` comes from `MonitorConn`, as you can see in the code. `MonitorConn` may close it, which will kill the connection. At this point we have an error that we would like to report ("connection was closed because ..."). But we cannot report it, since we are using a context which is already done. 

The correct thing is to use `cancelCtx` here, which has the correct scope.

Is this confusing in any way?

Reply:
I don't think user navigating anywhere has any impact here. As far as I know, this is fairly removed and the communication happens by means of objects stored in the backend.

Reply:
> ctx is later overwritten by MonitorConn.

That's correct but IIRC ConnectionDiagnosticID is only used by the web UI and:
- ssh connection is short-lived so it won't trigger the idle connection
- cert is generated and valid for 1 minute so it won't trigger the cert expiration error.

Reply:
@marcoandredinis thoughts?

Reply:
> cert is generated and valid for 1 minute so it won't trigger the cert expiration error.

Why not?

Personally I would prefer to think `MonitorConn` as a black box and it kills the connections for the reasons it cares:

https://github.com/gravitational/teleport/blob/0bf204cd6be950ddb67a9a1978ff3f29f4e2a0cc/lib/srv/db/proxyserver.go#L75-L79

I think the question is if the connection is no longer authorized, do we care to complete the diagnostics? If yes, then use `cancelCtx`, if not use `ctx`.

Reply:
We care about complete diagnostics :+1:

Whatever error happens, we should be able to present it to the user

Reply:
Thank you for all of your input, everyone. I'm going to merge this now.